### DISCUSSION TITLE
How to impliment IVector<T> interface that impliments INumber<T>?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [dotnet/runtime#90061] [Q&A]
How would one go about implementing a IVector<T> interface? 

I have ran into many problems while doing so.

For example I could not figure out how to do something like this...
```
public interface IVectorMultiplyOperators <TSelf, T> :
     IMultiplyOperators<TSelf, TSelf, TSelf>,
     IMultiplyOperators<T    , TSelf, TSelf>,
     IMultiplyOperators<TSelf, T    , TSelf>
{

}```

since you get this error...
https://cdn.discordapp.com/attachments/1137169841720131674/1137169841938239619/image.png

There is also the problem of implementing Create and Convert From and To methods
As well as Parsing and Formatting.

The reason I would like to use INumber as a base is because it holds a interface for some things like...
- One
- Zero
- Abs
- Min
- Max
- Sign
- Clamp

which only exists in INumber<T> and I don't think we can implement a custom interface for existing numeric structs

**Comment by tannergooding** [comment]
You wouldn't define all the different versions of `IMultiplyOperators`

You'd likely define something like this instead:
```csharp
public interface IVectorMultiplyOperators<TSelf, T> : IMultiplyOperators<TSelf, TSelf, TSelf>
    where TSelf : IVectorMultiplyOperators<TSelf, T>
    where T : INumber<T>
{
    static abstract TSelf operator *(TSelf left, T right);
}
```

Users who want to multiply a vector by a scalar would then need to constrain to `IVectorMultiplyOperators` as its a specialized operation specific to vectors and not a more general purpose operation that you'd encounter on more typical types.

Typical types are `T in, T out` and if a user has a `U` they need to first convert it to a `T`.

**Reply by tannergooding** [reply]
Otherwise, you end up with issues because `TSelf` and `T` could technically be the same and then `IMultiplyOperators<T    , TSelf, TSelf>`, `IMultiplyOperators<TSelf, T    , TSelf>`, and `IMultiplyOperators<TSelf, TSelf, TSelf>` all conflcit.

However, with the approach I gave above, you can have one method that implements both, since they can be disambiguated by full name (one is `IVectorMultiplyOperators.operator *` and the other `IMultiplyOperators.operator *`).

**Reply by Shadowblitz16** [reply]
The problem with that is there would be no way to implement both something like
Math.Mean(vector2a, vector2b);
Math.Mean(vector2a, numberb);

I want to implement a generic Math class over generic numbers and generic vectors and matricies

**Reply by tannergooding** [reply]
Due to how overload resolution, particularly with generics works; this won't be readily possible.

You cannot define an `IOperators<T, T>` with another `IOperators<T, U>` because `T` and `U` may be the same thing, which then leads to an ambiguity in the method table. That's also why defining an overload `operator +(T left, U right)` and its corresponding commutative operator `(operator +(U left, T right)` is problematic.
